bftengine/src/bftengine/KeyManager.cpp

```diff
@@ -0,0 +1,95 @@
+#include "KeyManager.h"
+#include "thread"
+#include "ReplicaImp.hpp"
+#include "ReplicaConfig.hpp"
+
+// KeyManager::KeyExchangeMsg::KeyExchangeMsg(KeyManager::KeyExchangeMsg&&
+// other):key_(std::move(other.key_)),signature_(other.signature_),repID_(other.repID_){}
+
+const std::string KeyManager::KeyExchangeMsg::getVersion() const { return "1"; }
+
+KeyManager::KeyExchangeMsg KeyManager::KeyExchangeMsg::KeyExDes(const char* serStr, const uint32_t& size) {
+  std::stringstream ss;
+  KeyManager::KeyExchangeMsg ke;
+  ss.write(serStr, std::streamsize(size));
+  KeyManager::KeyExchangeMsg::deserialize(ss, ke);
+  return ke;
+}
+
+void KeyManager::KeyExchangeMsg::serializeDataMembers(std::ostream& outStream) const {
+  serialize(outStream, key_);
+  LOG_TRACE(GL, "KEY EXCHANGE MANAGER  ser  key_ " << key_);
+  serialize(outStream, signature_);
+  LOG_TRACE(GL, "KEY EXCHANGE MANAGER  ser  signature_ " << signature_);
+  serialize(outStream, repID_);
+  LOG_TRACE(GL, "KEY EXCHANGE MANAGER  ser  repID_ " << repID_);
+}
+
+void KeyManager::KeyExchangeMsg::deserializeDataMembers(std::istream& inStream) {
+  deserialize(inStream, key_);
+  deserialize(inStream, signature_);
+  deserialize(inStream, repID_);
+}
+
+std::string KeyManager::KeyExchangeMsg::toString() const {
+  std::stringstream ss;
+  ss << "key [" << key_ << "] signature [" << signature_ << "] replica id [" << repID_ << "]";
+  return ss.str();
+}
+
+KeyManager& KeyManager::get() {
+  static KeyManager km;
+  return km;
+}
+void KeyManager::set(concord::kvbc::IClient* cl, const int& id) {
+  cl_ = cl;
+  repID_ = id;
+}
+
+void KeyManager::start(const int& durMs) {
```

Comment:
Can we make the variable name more readable?
Why is it `const int&`?
Can we pass a specific type aka `std::chrono::milliseconds`?

Reply:
I think it's an easier interface to pass an int rather than the chrono type.

Reply:
but I should mention that this class is a template class for @toly-kournik  to implement the real functionality.
This method will not be in production, but for testing purposes.

Reply:
The std::chrono types are much safer to use, and prevent a lot of conversion errors. I would prefer if we could use them for everything where time is relevant. Effort went into refactoring the timers and other parts of the code base to use std::chrono already, and not using them for convenience doesn't seem like a solid justification. We did in fact have bugs due to improper type conversion between seconds and milliseconds without them and they were very hard to find.